### PR TITLE
Implement Chain with Fuses

### PR DESCRIPTION
**Pull request by cuviper**
This change was inspired by the [proposal](https://internals.rust-lang.org/t/proposal-implement-iter-chain-using-fuse/12006) on the internals forum.

By implementing `Chain` with `Fuse`s, it no longer has to juggle the exhaustion state of each of its iterators, greatly simplifying the implementation. However, there was some concern that nested chains could wastefully recurse into exhausted sides when `FusedIterator` specialization is involved, so that possibility is hidden by a simple `Defuse` wrapper.

r? @scottmcm 

(edit: the original "Fuse with Option" part of this PR was merged in #70366.)

### CODE REVIEW

File: src/libcore/iter/adapters/chain.rs [hunk 1]
```diff
@@ -86,112 +54,61 @@ where
...
+        for x in self.a.by_ref() {
+            if n == 0 {
+                return Some(x);
             }
-            ChainState::Back => {}
+            n -= 1;
         }
-        if let ChainState::Back = self.state { self.b.nth(n) } else { None }
+        self.b.nth(n)
     }
```
Comment (KrishnaSannasi) [inline src/libcore/iter/adapters/chain.rs]:
```suggestion
    #[inline]
    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        let result_of_a = self.a.try_fold(n, |n, x| n.checked_sub(1).ok_or(x));
        match result_of_a {
            Ok(n) => self.b.nth(n),
            Err(item) => Some(item)
        }
    }
```

File: src/libcore/iter/adapters/chain.rs [hunk 2]
```diff
@@ -203,36 +120,32 @@ where
...
+
+    #[inline]
+    fn rfind<P>(&mut self, mut predicate: P) -> Option<Self::Item>
+    where
+        P: FnMut(&Self::Item) -> bool,
+    {
+        match self.b.rfind(&mut predicate) {
+            None => self.a.rfind(predicate),
+            item => item,
         }
```
Comment (KrishnaSannasi) [inline src/libcore/iter/adapters/chain.rs]:
```suggestion
    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        let result_of_b = self.b.try_rfold(n, |n, x| n.checked_sub(1).ok_or(x));
        match result_of_b {
            Ok(n) => self.a.nth_back(n),
            Err(item) => Some(item)
        }
    }
```

**Review by RustyYato — CHANGES_REQUESTED** [review]
`nth` and `nth_back` can be improved by avoiding the for loop

**Comment by cuviper** [comment]
@KrishnaSannasi 
> `nth` and `nth_back` can be improved by avoiding the for loop

I agree that's a good idea, but maybe it should be a followup PR? For the moment, I've kept mostly the same style of computation, apart from the direct simplification in state tracking.

FWIW, the internal `LoopState` might be a better type for your implementation. I also try to avoid closures in iterator implementations (#62429), but that one would be an easy generic function, no captured values needed.

**Comment by KrishnaSannasi** [comment]
@cuviper That sounds like a good idea, can I make a PR after if this one is merged?

**Comment by cuviper** [comment]
Sure, it's all yours.

**Comment by cuviper** [comment]
1188.7% change in deeply-nested-opt!?!

**Comment by scottmcm** [comment]
Oh, that benchmark is this:
```rust
    Box::new(empty()
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty()) // 10th .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty())
        .chain(empty()) // 16th .chain(empty())
    )
```

I guess it makes sense that if this change is going to impact anything, it'd be that 😕

**Comment by scottmcm** [comment]
@cuviper I suppose this _is_ making two changes at once.  Your idea to do `Fuse` with `Option` seems clearly-good and separable from this; want to make a PR (either this one or a new one) with just that part?  That might help make it easier to isolate what's going on in perf.

**Comment by cuviper** [comment]
I was extra confused for a bit that it looks like that benchmark only tests construction, but it returns a `Box<Iterator>`, so it will have to codegen the entire `Iterator` vtable...

**Comment by cuviper** [comment]
That perf test is _evil_, but I'm a little inclined to accept the loss here. The performance here is not as bad as the original issue #38528, and I found alternate cases that are much worse on the current compiler, but OK with this PR.

The original performance problem was up to Rust 1.25, where I get 11.37s for `build --release`. With the fix #48296 in Rust 1.26, I get 0.24s, and on `1.44.0-nightly (537ccdf3a 2020-04-02)` I get 0.20s. With this `Chain` PR I get 1.37s -- definitely worse, but not a full regression.

However, `Box<Iterator<Item = ()>>` is pretty useless, and your comment about ZSTs made me think to try other `Item` types (still with `Empty`). With `bool`, nightly gets 0.26s and this PR gets 1.43s. With `NonZeroU32`, nightly gets 0.24s and this PR gets 1.40s. With `u32`, nightly **did not finish** (burned 200% CPU for 5 minutes before I killed it), yet this PR gets 1.43s.

I'll file a new issue for that -- something in the compiler needs to be fixed regardless of whether we change `Chain`.

---

The rest of the perf report still showed some slowdowns, though not as drastic, so I'll also explore if there are any other tweaks I can make here.

**Comment by cuviper** [comment]
> The original performance problem was up to Rust 1.25, where I get 11.37s for `build --release`. With the fix #48296 in Rust 1.26, I get 0.24s, and on `1.44.0-nightly (537ccdf3a 2020-04-02)` I get 0.20s. With this `Chain` PR I get 1.37s -- definitely worse, but not a full regression.

When rebased onto #70750, this PR is down to 0.45s compiling deeply-nested. So the added iterator layers still slow us down, but maybe it's close enough to write off for this pathological test.

> With `u32`, nightly **did not finish** (burned 200% CPU for 5 minutes before I killed it), yet this PR gets 1.43s.

See issue #70749 for the hang. With #70750, this PR gets down to 0.54s.

**Comment by cuviper** [comment]
Let's measure again now that #70750 has landed.

@bors try @rust-timer queue

**Comment by CAD97** [comment]
- deeply-nested-opt: 124.7%
- deeply-nested-debug: 28.3%
- otherwise, close to noise (but trending slower)

**Comment by cuviper** [comment]
I've posted an alternate in #70896 -- the code is not as pretty as the free `Fuse` would get us here, but it seems to perform a lot better.

**Comment by cuviper** [comment]
#70896 has merged, so I'm closing this.

**Pull request closed without merging**